File: docs/vsto/codesnippet/CSharp/Trin_Outlook_FR_Import/TaskFormRegion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Resources;
using System.Text;
using Office = Microsoft.Office.Core;
using Outlook = Microsoft.Office.Interop.Outlook;
using OutlookTools = Microsoft.Office.Tools.Outlook;

namespace Trin_Outlook_FR_Import
{
    public partial class TaskFormRegion
    {
        #region Form Region Factory

        [Microsoft.Office.Tools.Outlook.FormRegionMessageClass("IPM.Task.TaskFormRegion")]
        [Microsoft.Office.Tools.Outlook.FormRegionName("Trin_Outlook_FR_Import.TaskFormRegion")]
        public partial class TaskFormRegionFactory
        {
            private void InitializeManifest()
            {
                ResourceManager resources = new ResourceManager(typeof(TaskFormRegion));
                this.Manifest.FormRegionType = OutlookTools.FormRegionType.ReplaceAll;
                this.Manifest.Title = resources.GetString("Title");
                this.Manifest.FormRegionName = resources.GetString("FormRegionName");
                this.Manifest.Description = resources.GetString("Description");
                this.Manifest.ShowInspectorCompose = true;
                this.Manifest.ShowInspectorRead = true;
                this.Manifest.ShowReadingPane = true;

            }

            // Occurs before the form region is initialized.
            // To prevent the form region from appearing, set e.Cancel to true.
            // Use e.OutlookItem to get a reference to the current Outlook item.
            private void TaskFormRegionFactory_FormRegionInitializing(object sender, OutlookTools.FormRegionInitializingEventArgs e)
            {
            }
        }

        #endregion

        // Occurs before the form region is displayed.
        // Use this.OutlookItem to get a reference to the current Outlook item.
        // Use this.OutlookFormRegion to get a reference to the form region.
        //<Snippet5>
        private void TaskFormRegion_FormRegionShowing
            (object sender, EventArgs e)
        {
            populateComboBox();
            RefreshTaskListBox();
            this.addDependentTask.Click += new
            Microsoft.Office.Interop.Outlook.
                OlkCommandButtonEvents_ClickEventHandler(AddDependentTask_Click);

        }
        //</Snippet5>
        //<Snippet1>
        private void populateComboBox()
        {
            Outlook.Application Application = new Outlook.Application();
            Outlook.NameSpace outlookNameSpace = Application.GetNamespace("MAPI");
            Outlook.MAPIFolder taskFolder = outlookNameSpace.GetDefaultFolder(
                Microsoft.Office.Interop.Outlook.OlDefaultFolders.olFolderTasks);
            Outlook.Items taskItems = taskFolder.Items;
            foreach (Outlook.TaskItem task in taskItems)
            {
                if (task.Subject != null)
                {
                    comboBox1.AddItem(task.Subject, System.Type.Missing);
                }
            }
            comboBox1.Text = comboBox1.GetItem(0);
        }
        //</Snippet1>
        //<Snippet2>
        void AddDependentTask_Click()
        {
            Outlook.TaskItem tempTaskItem = FindTaskBySubjectName(comboBox1.Text);
            if (tempTaskItem != null)
            {
                this.listBox1.AddItem(tempTaskItem.PercentComplete.ToString()
    + "% Complete -- " + tempTaskItem.Subject, System.Type.Missing);
                this.olkTextBox3.Text = this.olkTextBox3.Text + "|" + 
                    tempTaskItem.Subject;
            }
        }
        //</Snippet2>
        //<Snippet3>
        private Outlook.TaskItem FindTaskBySubjectName(string subjectName)
        {
            Outlook.Application Application = new Outlook.Application();
            Outlook.NameSpace outlookNameSpace = Application.GetNamespace("MAPI");
            Outlook.MAPIFolder tasksFolder =
                outlookNameSpace.GetDefaultFolder(
            Microsoft.Office.Interop.Outlook.
                OlDefaultFolders.olFolderTasks);
            Outlook.Items taskItems = tasksFolder.Items;
            foreach (Outlook.TaskItem taskItem in taskItems)
            {
                if (taskItem.Subject == subjectName)
                {
                    return taskItem;
                }
            }
            return null;
        }
        //</Snippet3>
        //<Snippet4>
        void RefreshTaskListBox()
        {
            listBox1.Clear();
            Outlook.TaskItem tempTaskItem;
            String[] tempArray = olkTextBox3.Text.Split(new Char[] { '|' });

            foreach (string tempString in tempArray)
            {
                tempTaskItem = FindTaskBySubjectName(tempString);
                if (tempTaskItem != null)
                {
                    this.listBox1.AddItem(tempTaskItem.PercentComplete.ToString()
                        + "% Complete -- " + tempTaskItem.Subject, 
                            System.Type.Missing);
                }
            }
        }
        //</Snippet4>

        // Occurs when the form region is closed.
        // Use this.OutlookItem to get a reference to the current Outlook item.
        // Use this.OutlookFormRegion to get a reference to the form region.
        private void TaskFormRegion_FormRegionClosed(object sender, 
            System.EventArgs e)
        {
        }
    }
}

File: docs/vsto/codesnippet/VisualBasic/Trin_Outlook_FR_Import_O12/TaskFormRegion.vb
Public Class TaskFormRegion

#Region "Form Region Factory"

    <Microsoft.Office.Tools.Outlook.FormRegionMessageClass("IPM.Task.TaskFormRegion")> _
    <Microsoft.Office.Tools.Outlook.FormRegionName("Trin_Outlook_FR_Import_O12.TaskFormRegion")> _
    Partial Public Class TaskFormRegionFactory

    Private Sub InitializeManifest()
            Dim resources As System.Resources.ResourceManager = New System.Resources.ResourceManager(GetType(TaskFormRegion))
            Me.Manifest.FormRegionType = Microsoft.Office.Tools.Outlook.FormRegionType.ReplaceAll
            Me.Manifest.Title = resources.GetString("Title")
            Me.Manifest.FormRegionName = resources.GetString("FormRegionName")
            Me.Manifest.Description = resources.GetString("Description")
            Me.Manifest.ShowInspectorCompose = True
            Me.Manifest.ShowInspectorRead = True
            Me.Manifest.ShowReadingPane = True

    End Sub

    ' Occurs before the form region is initialized.
    ' To prevent the form region from appearing, set e.Cancel to true.
    ' Use e.OutlookItem to get a reference to the current Outlook item.
        Private Sub TaskFormRegionFactory_FormRegionInitializing(ByVal sender As Object, ByVal e As Microsoft.Office.Tools.Outlook.FormRegionInitializingEventArgs) Handles Me.FormRegionInitializing

    End Sub

    End Class

#End Region
    ' Occurs before the form region is displayed.
    ' Use this.OutlookItem to get a reference to the current Outlook item.
    ' Use this.OutlookFormRegion to get a reference to the form region.
    '<Snippet5>
    Private Sub TaskFormRegion_FormRegionShowing(ByVal sender As Object, _
        ByVal e As System.EventArgs) Handles MyBase.FormRegionShowing
        populateComboBox()
        RefreshTaskListBox()
    End Sub
    '</Snippet5>
    ' Occurs when the form region is closed.
    ' Use this.OutlookItem to get a reference to the current Outlook item.
    ' Use this.OutlookFormRegion to get a reference to the form region.
    Private Sub TaskFormRegion_FormRegionClosed(ByVal sender As Object, _
        ByVal e As System.EventArgs) Handles MyBase.FormRegionClosed

    End Sub
    '<Snippet1>
    Private Sub populateComboBox()
        Dim Application As New Outlook.Application()
        Dim outlookNameSpace As Outlook.NameSpace = _
        Application.GetNamespace("MAPI")
        Dim taskFolder As Outlook.MAPIFolder = _
        outlookNameSpace.GetDefaultFolder _
            (Microsoft.Office.Interop.Outlook.OlDefaultFolders.olFolderTasks)
        Dim taskItems As Outlook.Items = taskFolder.Items
        Dim task As Outlook.TaskItem
        For Each task In taskItems
            If Not (task.Subject Is Nothing) Then
                comboBox1.AddItem(task.Subject, System.Type.Missing)
            End If
        Next task
        comboBox1.Text = comboBox1.GetItem(0)
    End Sub
    '</Snippet1>
    '<Snippet2>
    Private Sub AddDependentTask_Click1() Handles addDependentTask.Click
        Dim tempTaskItem As Outlook.TaskItem = FindTaskBySubjectName _
            (comboBox1.Text)
        If Not (tempTaskItem Is Nothing) Then
            Me.listBox1.AddItem(tempTaskItem.PercentComplete.ToString() & _
                "% Complete -- " & tempTaskItem.Subject, System.Type.Missing)
            Me.olkTextBox3.Text = Me.olkTextBox3.Text & "|" & tempTaskItem.Subject
        End If
    End Sub
    '</Snippet2>
    '<Snippet3>
    Private Function FindTaskBySubjectName(ByVal subjectName As String) _
        As Outlook.TaskItem
        Dim Application As New Outlook.Application()
        Dim outlookNameSpace As Outlook.NameSpace = _
            Application.GetNamespace("MAPI")
        Dim tasksFolder As Outlook.MAPIFolder = _
            outlookNameSpace.GetDefaultFolder _
                (Microsoft.Office.Interop.Outlook.OlDefaultFolders.olFolderTasks)
        Dim taskItems As Outlook.Items = tasksFolder.Items
        Dim taskItem As Outlook.TaskItem
        For Each taskItem In taskItems
            If taskItem.Subject = subjectName Then
                Return taskItem
            End If
        Next taskItem
        Return Nothing
    End Function
    '</Snippet3>
    '<Snippet4>
    Private Sub RefreshTaskListBox()
        listBox1.Clear()
        Dim tempTaskItem As Outlook.TaskItem
        Dim tempArray As [String]() = olkTextBox3.Text.Split(New [Char]() {"|"c})

        Dim tempString As String
        For Each tempString In tempArray
            tempTaskItem = FindTaskBySubjectName(tempString)
            If Not (tempTaskItem Is Nothing) Then
                Me.listBox1.AddItem(tempTaskItem.PercentComplete.ToString() & _
                    "% Complete -- " & tempTaskItem.Subject, System.Type.Missing)
            End If
        Next tempString
    End Sub
    '</Snippet4>

End Class
